### PR TITLE
Support multi-batch triangle attention

### PR DESCRIPTION
**Pull request by wujingyue**
## Summary
- flatten multiple batch dims in sdpa forward so we can reuse flash attention kernels and undo the flattening for outputs
- update logical domain mapping to match the new sdpa producer/consumer layout and avoid mapping philox outputs
- tighten dynamic transform concretizer iteration and update AlphaFold3 test to call sdpa with explicit is_causal flag

cc @DejunL

### CODE REVIEW

**Comment by wujingyue** [comment]
!test

File: csrc/logical_domain_map.cpp [hunk 1]
```diff
@@ -361,35 +361,44 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseLogicalDomainMap::map(
...
     }
-    size_t num_device_dim = producer_logical.at(0)->isDeviceDim() ? 1 : 0;
-    // Map N, H from any input (query/key/value)
-    for (auto idx : arange(consumer_root.size())) {
-      if (idx < (2 + num_device_dim)) {
-        updatePairwiseLogicalDomainMap(
-            producer_logical.at(idx), consumer_root.at(idx));
+
+    const auto num_batch_dims =
+        std::ssize(op->attn_out()->getLogicalDomain()) - 2;
```
Comment (xwang233) [inline csrc/logical_domain_map.cpp:388]:
```suggestion
        static_cast<std::size_t>(std::ssize(op->attn_out()->getLogicalDomain()) - 2);
```
**[Suggested by AI]** Summary of the fix

Problem  
• std::ssize(...) returns a signed type (typically std::ptrdiff_t).  
• The result was later being used where an unsigned type (std::size_t) was expected, which produced a build/​conversion error or warning.

Change made  
• Wrapped the result of the expression in a static_cast to std::size_t:

    static_cast<std::size_t>(std::ssize(op->attn_out()->getLogicalDomain()) - 2);

Effect  
• Forces an explicit, well-defined conversion from the signed value to an unsigned size_t.  
• Eliminates the compiler error/warning about implicit conversion or type mismatch, allowing the code to build successfully.

**Comment by wujingyue** [comment]
!test

File: tests/python/test_alphafold3.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+        b_h = fd.ops.linear(z_in, w_b)
+
+        v = fd.ops.linear(z_in, w_v)
+        v_h = fd.ops.reshape(
+            v, [batch_size, n_tokens, n_tokens, h, -1]
+        )  # [b, i, j, h, c_hidden]
+        v_h = fd.ops.permute(v_h, [0, 1, 3, 2, 4])  # [b, i, h, j, c_hidden]
+
+        # TODO: b_h should be added here. fd.ops.sdpfa_fwd hasn't yet supported custom masks.
```
Comment (wujingyue) [inline tests/python/test_alphafold3.py]: @Priya2698, could SdpaFwdOp call https://docs.pytorch.org/cppdocs/api/function_namespaceat_1a32d6168f7e2945664bd5e565a14084e2.html#exhale-function-namespaceat-1a32d6168f7e2945664bd5e565a14084e2 instead, which supports attn_mask?

File: tests/python/test_alphafold3.py
Comment (wujingyue) [inline tests/python/test_alphafold3.py:1]: @DejunL to review this file. It contains only single-GPU implementation as a baseline.

**Comment by wujingyue** [comment]
!test

File: tests/python/test_alphafold3.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        o_h, _, _, _ = fd.ops.sdpfa_fwd(
+            q_h, k_h, v_h, is_causal=False
+        )  # [b, i, h, j, c_hidden]
```
Comment (DejunL) [inline tests/python/test_alphafold3.py:92]: without knowing the underlying sdpfa_fwd() logics, it seems this is missing the triangle bias and padding mask processing, both of which are needed in the triangle attention CUDA kernel [here](https://docs.nvidia.com/cuda/cuequivariance/api/generated/cuequivariance_torch.triangle_attention.html)
-- I see your inline comments above but I still don't see the padding mask in this workflow

File: tests/python/test_alphafold3.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+# https://elanapearl.github.io/blog/2024/the-illustrated-alphafold/#triangle-attention
+def test_triangle_attention_starting_node():
+    c_z, c_hidden, h = (
+        _DEFAULT_CONFIG.c_z,
+        _DEFAULT_CONFIG.c_hidden,
+        _DEFAULT_CONFIG.n_heads,
+    )
+
+    with FusionDefinition() as fd:
```
Comment (DejunL) [inline tests/python/test_alphafold3.py:41]: Just curious, is this whole block under the contextmanager what the user needs to do in order to trigger the JIT fusion? If not, as I'm not familar with the nvFuser API, it would be helpful to provide the actual usage example here

File: tests/python/test_alphafold3.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        z_in = fd.define_tensor(
```
Comment (DejunL) [inline tests/python/test_alphafold3.py:42]: Before the q/k/v projection from z, there's an extra `z = layernorm(x)` in the AF3 definition of triangle attention. See [here](https://gitlab-master.nvidia.com/clara-discovery/boltz/-/blob/dev/src/boltz/distributed/model/layers/triangular_attention.py?ref_type=heads#L1530)

Comment (wujingyue) [reply tests/python/test_alphafold3.py:42]: Yes, I understood a layernorm should happen before linear projections. I excluded layernorm from this test because
1. Layernorm is not in https://elanapearl.github.io/blog/2024/the-illustrated-alphafold/#triangle-attention but I'm sure people define what triangle attention includes differently. 
2. I wanted tests in this file to focus on the hardest part for nvFuser. layernorm doesn't seem challenging -- it's per token and doesn't involve communication. 
3. layernorm is a bit cumbersome to write in nvFuser's Python API -- it is not an fd.ops but [a series of them](https://github.com/NVIDIA/Fuser/blob/7f921200fbc25e70afa45990bbd373eee44d34db/tests/python/multidevice/test_transformer.py#L223-L246).

File: tests/python/test_alphafold3.py (hunk 4 above)
Comment (wujingyue) [reply tests/python/test_alphafold3.py:41]: I should have clarified the context much better. 

> is this whole block under the contextmanager what the user needs to do in order to trigger the JIT fusion?

No, the user doesn't need to program in nvFuser. The actual usage is more like https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/tests/distributed/test_dtensor.py#L291. The rough order is
1. writes a single-GPU torch.nn.Module,
2. `parallelize_module` that module indicating how parameters are sharded using ParallelStyle (like ColwiseParallel)
3. `thunder.jit` that parallelized module,
4. run that JIT'ed module on DTensors. 

These tests are written in nvFuser so "frontend" and "backend" can be worked on in parallel. The Thunder/torch.compile team focuses on generating FusionDefinitions and the nvFuser team focuses on executing FusionDefinitions.

File: tests/python/test_alphafold3.py (hunk 3 above)
Comment (wujingyue) [reply tests/python/test_alphafold3.py:92]: Great to know this! I wasn't sure if the mask is needed because http://nv/e(O is optional. I'll add that as an extra input. However, the same limitation of fd.ops.sdpfa_fwd applies and should be fixed. 

Question: are the mask and the triangle bias additive? I.e. `attention_scores = softmax((q @ k.T / sqrt(d)) + (mask == 1 ? 0 : -inf) + triangle_bias)`?

Comment (wujingyue) [reply tests/python/test_alphafold3.py:92]: We need an SDPA kernel that takes `mask` and `bias` **separately**.[^1] `mask` and `bias` have different broadcast dimensions -- `mask` has shape `[B,N,1,1,K]` and `bias` has shape `[B,1,H,Q,K]`. While we could combine the two to one mask tensor to reuse PyTorch's SDPA kernel, that combined mask tensor would have to be of shape `[B,N,H,Q,K]` and therefore too large to materialize. 

[^1]: @DejunL reminded me of this important detail the first time we talk.

Comment (wujingyue) [reply tests/python/test_alphafold3.py:92]: Question for @DejunL: what does the mask look like? If it's causal, https://docs.pytorch.org/cppdocs/api/function_namespaceat_1a32d6168f7e2945664bd5e565a14084e2.html#exhale-function-namespaceat-1a32d6168f7e2945664bd5e565a14084e2 could be used. Otherwise, we can consider adding cuequivariance_torch as a dependency.

Comment (DejunL) [reply tests/python/test_alphafold3.py:92]: > Question: are the mask and the triangle bias additive? I.e. attention_scores = softmax((q @ k.T / sqrt(d)) + (mask == 1 ? 0 : -inf) + triangle_bias)?

Both mask_bias and triangle bias are added before softmax: 
```python
mask_bias = mask == 1 ? 0 : -inf # e.g., inf = 1e9 for FP32
attention_scores = softmax(q @ k.T / sqrt(d) + mask_bias + triangle_bias)
``` 
See our reference implementation [here](https://gitlab-master.nvidia.com/clara-discovery/boltz/-/blob/dev/src/boltz/model/layers/triangular_attention/primitives.py?ref_type=heads#L457)

> what does the mask look like?

In our cuequivariance implementation, we support completely random mask, i.e., there is no deterministic pattern of 1 vs 0 in it. On the other hand, in most practical usage cases, it's square mask where `mask[i_batch, : n_valid_tokens_this_sample, : n_valid_tokens_this_sample] == 1` but otherwise zeros, i.e., only a sub-square towards the upper left corner of each sample is 1 but 0 everywhere else

> While we could combine the two to one mask tensor to reuse PyTorch's SDPA kernel, that combined mask tensor would have to be of shape [B,N,H,Q,K] and therefore too large to materialize

Yes, and therefore we need the cueq TriAttn kernel for this

**Review by DejunL — APPROVED** [review]
LGTM

File: csrc/logical_domain_map.cpp [hunk 6]
```diff
@@ -361,35 +361,44 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseLogicalDomainMap::map(
...
+    //   value = [N*, S, Ev]
+    //
     // Consumers:
-    //   output = [DIDx(D)?, N, H, L, Ev]
-    //   logsumexp = [DIDx(D)?, N, H, L]
-    // Note: S, E are not mapped together in the producers and do not have any
+    //   output = [N*, L, Ev]
+    //   logsumexp = [N*, L]
+    //   philox_seed = []
+    //   philox_offset = []
```
Comment (Priya2698) [inline csrc/logical_domain_map.cpp:373]: The shape of philox_seed is different based on version: https://github.com/NVIDIA/Fuser/blob/f28e6d8c7057ce5de9d22c9a9924569df1e1305e/csrc/ir/internal_nodes.h#L2507C39-L2514

I recommend either specifying the shapes version-wise or skipping them entirely as they are irrelevant to logical domain mapping.

File: csrc/logical_domain_map.cpp [hunk 7]
```diff
@@ -361,35 +361,44 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseLogicalDomainMap::map(
   }
 
   if (auto* op = dynamic_cast<SdpaFwdOp*>(consumer_tv_->definition())) {
-    // Note: Explicit handling of DIDx(D) until
-    // https://github.com/NVIDIA/Fuser/issues/2563 is resolved. Producers:
-    //   query = [DIDx(D)?, N, H, L, E]
-    //   key = [DIDx(D)?, N, H, S, E]
-    //   value = [DIDx(D)?, N, H, S, Ev]
+    // Producers:
+    //   query = [N*, L, E]
```
Comment (Priya2698) [inline csrc/logical_domain_map.cpp:365]: Why is the head dimension collapsed into batch as well? Is this just for the ease of mapping?

File: csrc/logical_domain_map.cpp [hunk 8]
```diff
@@ -361,35 +361,44 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseLogicalDomainMap::map(
...
     // Consumers:
-    //   output = [DIDx(D)?, N, H, L, Ev]
-    //   logsumexp = [DIDx(D)?, N, H, L]
-    // Note: S, E are not mapped together in the producers and do not have any
+    //   output = [N*, L, Ev]
+    //   logsumexp = [N*, L]
+    //   philox_seed = []
+    //   philox_offset = []
+    //
+    // N* are the dimensions that are treated batch, e.g., the actual batch
```
Comment (Priya2698) [inline csrc/logical_domain_map.cpp:375]: This sounds confusing to me. In the SDPA code above, the head dimension is not collapsed into batch.

File: tests/python/test_alphafold3.py (hunk 2 above)
Comment (Priya2698) [reply tests/python/test_alphafold3.py]: Currently, nvfuser API is registered against flash attention in Thunder so we cannot simply swap the backend out.

To support other attention variants, we can expand the API and use `_fused_sdp_choice` to dispatch to different variants (https://github.com/pytorch/pytorch/blob/2887faaec6295d081580d09fce161201826c6d87/aten/src/ATen/native/transformers/attention.cpp#L738) within `SdpaFwdOp`. This will require some changes in how we register nvfuser sdpa op in Thunder as well. 
We cannot use `scaled_dot_product_attention` as it does not return enough information for backward pass.

File: csrc/logical_domain_map.cpp (hunk 6 above)
Comment (wujingyue) [reply csrc/logical_domain_map.cpp:373]: > skipping them entirely as they are irrelevant to logical domain mapping.

I believe that's what I did: https://github.com/NVIDIA/Fuser/pull/5597/files/e6e4c5428cb580f31d4f1b851cc985b10d674144#diff-2c6b12531ecac398182531099318537e3708ba13f674bfa0c42b7b679013d1faR382-R385

Do you mean to remove philox_seed and philox_offset from the comment? Yes, I can certainly do that to avoid confusion.

File: csrc/logical_domain_map.cpp (hunk 7 above)
Comment (wujingyue) [reply csrc/logical_domain_map.cpp:365]: N* is a **list** of dimensions that can be treated as batch dimensions for the mapping purpose. So the head dimension is not **collapsed** into batch.

File: csrc/logical_domain_map.cpp (hunk 8 above)
Comment (wujingyue) [reply csrc/logical_domain_map.cpp:375]: ditto

File: csrc/logical_domain_map.cpp (hunk 6 above)
Comment (Priya2698) [reply csrc/logical_domain_map.cpp:373]: > Do you mean to remove philox_seed and philox_offset from the comment? Yes, I can certainly do that to avoid confusion.

Yes

**Comment by wujingyue** [comment]
!test
